app/gui2/src/components/GraphEditor/widgets/WidgetFunctionName.vue

```diff
@@ -0,0 +1,97 @@
+<script setup lang="ts">
+import AutoSizedInput from '@/components/widgets/AutoSizedInput.vue'
+import { Score, WidgetInput, defineWidget, widgetProps } from '@/providers/widgetRegistry'
+import { useProjectStore } from '@/stores/project'
+import type { SuggestionEntry } from '@/stores/suggestionDatabase/entry'
+import { Ast } from '@/util/ast'
+import { Err, Ok, type Result } from '@/util/data/result'
+import { useToast } from '@/util/toast'
+import { PropertyAccess } from 'shared/ast'
+import type { ExpressionId } from 'shared/languageServerTypes'
+import { computed, ref, watchEffect } from 'vue'
+import NodeWidget from '../NodeWidget.vue'
+
+const props = defineProps(widgetProps(widgetDefinition))
+const displayedName = ref(props.input.value.code())
+
+const project = useProjectStore()
+const renameError = useToast.error()
+
+const thisArg = computed(() =>
+  props.input.value instanceof PropertyAccess ? props.input.value.lhs : undefined,
+)
+const operator = computed(() =>
+  props.input.value instanceof PropertyAccess ? props.input.value.operator : undefined,
+)
+const name = computed(() =>
+  props.input.value instanceof PropertyAccess ? props.input.value.rhs : props.input.value,
+)
+watchEffect(() => (displayedName.value = name.value.code()))
+
+function newNameAccepted(name: string | undefined) {
+  if (!name) return
+  renameFunction(name).then((result) => renameError.reportError(result))
+}
+
+async function renameFunction(newName: string): Promise<Result> {
+  if (!project.modulePath?.ok) return project.modulePath ?? Err('Unknown module Path')
+  const refactorResult = await project.lsRpcConnection.renameSymbol(
+    project.modulePath.value,
+    props.input[FunctionName].editableName,
+    newName,
+  )
+  if (!refactorResult.ok) return refactorResult
+  displayedName.value = refactorResult.value.newName
```

Comment:
Not supre sure if it is OK to modify the local state after `await`. This introduces a race condition with other changes happening in the document. I think this should only be applied in case `displayedName` wasn't changed in the mean time, or not at all. After all, the name change should be propagated through document update anyway.